File: csharp/channel_partner/ImportChannelPartnerOrder.cs
using System;
using System.Collections.Generic;
using System.Reflection;
using com.ultracart.admin.v2.Api;
using com.ultracart.admin.v2.Model;

namespace SdkSample.channel_partner
{
    public class ImportChannelPartnerOrder
    {
        /*
            To run channel partner examples, you will need:
            1) An API Key: https://secure.ultracart.com/merchant/configuration/apiManagementApp.do
            2) That API Key must be assigned to a channel partner: https://secure.ultracart.com/merchant/configuration/customChannelPartnerListLoad.do

            The spreadsheet import docs will serve you well here.  They describe many fields
            https://ultracart.atlassian.net/wiki/spaces/ucdoc/pages/1377246/Channel+Partner+API+-+Spreadsheet+Import
        */
        public static void Execute()
        {
            Console.WriteLine("--- " + MethodBase.GetCurrentMethod()?.DeclaringType?.Name + " ---");
            var doFirstExample = false;
            try
            {
                // Create channel partner API instance using API key
                ChannelPartnerApi channelPartnerApi = new ChannelPartnerApi(Constants.ChannelPartnerApiKey);
                
                // NOTICE:
                // The real difficulty with using this API is the hosted fields requirement for credit card information.
                // You will need to incorporate UltraCart hosted fields in your Customer Service UI and capture credit card
                // information through the hosted fields and then provide the tokens here.  You CANNOT provide raw credit
                // card information via this interface.
                // The two fields in this API are hostedFieldsCardToken and hostedFieldsCvvToken
                // Within this sdk_samples github project, review the /hosted_fields/hosted_fields.html file for an example
                
                // There are TWO examples.  The first is an order that still needs the credit card charged.  This example uses
                // the hosted fields to collect payment information and pass it to UltraCart for processing.  The second example
                // already has payment processed and just passes in the authorization information.
                
                // ---------------------------------------------
                // ---------------------------------------------
                // Example 1 - Order needs payment processing
                // ---------------------------------------------
                // ---------------------------------------------
                
                ChannelPartnerOrder order = new ChannelPartnerOrder();

                if (doFirstExample)
                {

                    // order.AdvertisingSource = "Friend"; // https://ultracart.atlassian.net/wiki/spaces/ucdoc/pages/1377001/Advertising+Sources
                    // order.AffiliateId = 856234; // https://ultracart.atlassian.net/wiki/spaces/ucdoc/pages/1377727/Affiliates
                    // order.AffiliateSubId = 1234; // https://ultracart.atlassian.net/wiki/spaces/ucdoc/pages/1376754/Allowing+Affiliates+to+use+Sub-IDs
                    // order.ArbitraryShippingHandlingTotal = 9.99;
                    // order.ArbitraryTax = 2.50;
                    // order.ArbitraryTaxRate = 7.0;
                    // order.ArbitraryTaxableSubtotal = 69.99;

                    order.AssociateWithCustomerProfileIfPresent = true;
                    order.AutoApprovePurchaseOrder = true;
                    order.BilltoAddress1 = "11460 Johns Creek Parkway";
                    order.BilltoAddress2 = "Suite 101";
                    order.BilltoCity = "Duluth";
                    order.BilltoCompany = "Widgets Inc";
                    order.BilltoCountryCode = "US";
                    order.BilltoDayPhone = "6784153823";
                    order.BilltoEveningPhone = "6784154019";
                    order.BilltoFirstName = "John";
                    order.BilltoLastName = "Smith";
                    order.BilltoPostalCode = "30097";
                    order.BilltoStateRegion = "GA";
                    order.BilltoTitle = "Sir";
                    order.CcEmail = "orders@widgets.com";
                    order.ChannelPartnerOrderId = "widget-1245-abc";
                    order.ConsiderRecurring = false;
                    order.Coupons = new List<string> { "10OFF", "BUY1GET1" };

                    // order.CreditCardAuthorizationAmount = 69.99;
                    // order.CreditCardAuthorizationDts = DateTime.Now.ToString("yyyy-MM-ddTHH:mm:ssK"); // this will usually not be 'now'. it will be the actual auth date
                    // order.CreditCardAuthorizationNumber = "1234";

                    order.CreditCardExpirationMonth = 5;
                    order.CreditCardExpirationYear = 2032;
                    order.CreditCardType = "VISA"; // see the hosted fields below for the card number and cvv tokens
                    order.CustomField1 = "Whatever";
                    order.CustomField2 = "You";
                    order.CustomField3 = "Want";
                    order.CustomField4 = "Can";
                    order.CustomField5 = "Go";
                    order.CustomField6 = "In";
                    order.CustomField7 = "CustomFields";
                    order.DeliveryDate = DateTime.Now.ToString("yyyy-MM-ddTHH:mm:ssK");
                    order.Email = "ceo@widgets.com";
                    order.Gift = false;

                    order.GiftEmail = "sally@aol.com";
                    order.GiftMessage = "Congratulations on your promotion!";

                    order.HostedFieldsCardToken = "7C97B0AAA26AB10180B4B29F00380101";
                    order.HostedFieldsCvvToken = "C684AB4336787F0180B4B51971380101";

                    // order.InsuranceApplicationId = insuranceApplicationId; // these are used by only a handful of specialized merchants
                    // order.InsuranceClaimId = insuranceClaimId; // these are used by only a handful of specialized merchants

                    order.IpAddress = "34.125.95.217";

                    // -- Items start ---
                    ChannelPartnerOrderItem item = new ChannelPartnerOrderItem();
                    // item.ArbitraryUnitCost = 9.99;
                    // item.AutoOrderLastRebillDts = DateTime.Now.ToString("yyyy-MM-ddTHH:mm:ssK");
                    // item.AutoOrderSchedule = "Weekly";

                    item.MerchantItemId = "shirt";
                    item.Quantity = 1;
                    item.Upsell = false;

                    ChannelPartnerOrderItemOption itemOption1 = new ChannelPartnerOrderItemOption();
                    itemOption1.Name = "Size";
                    itemOption1.Value = "Small";

                    ChannelPartnerOrderItemOption itemOption2 = new ChannelPartnerOrderItemOption();
                    itemOption2.Name = "Color";
                    itemOption2.Value = "Orange";

                    item.Options = new List<ChannelPartnerOrderItemOption> { itemOption1, itemOption2 };

                    order.Items = new List<ChannelPartnerOrderItem> { item };
                    // -- Items End ---


                    order.LeastCostRoute = true; // Give me the lowest cost shipping
                    order.LeastCostRouteShippingMethods = new List<string>
                        { "FedEx: Ground", "UPS: Ground", "USPS: Priority" };
                    order.MailingListOptIn =
                        true; // Yes, I confirmed with the customer personally they wish to be on my mailing lists.
                    order.NoRealtimePaymentProcessing = false;
                    order.PaymentMethod = ChannelPartnerOrder.PaymentMethodEnum.CreditCard;
                    // order.PurchaseOrderNumber = "PO-12345";
                    order.RotatingTransactionGatewayCode =
                        "MyStripe"; // We wish this to be charged against our Stripe gateway
                    order.ScreenBrandingThemeCode =
                        "SF1986"; // Theme codes predated StoreFronts. Each StoreFront still has a theme code under the hood. We need that here. See this screen to find your code: https://secure.ultracart.com/merchant/configuration/customerServiceLoad.do
                    order.ShipOnDate = DateTime.Now.ToString("yyyy-MM-ddTHH:mm:ssK");
                    order.ShipToResidential = true;
                    // order.ShippingMethod = "FedEx: Ground"; // We're using LeastCostRoute, so we do not supply this value
                    order.ShiptoAddress1 = "55 Main Street";
                    order.ShiptoAddress2 = "Suite 202";
                    order.ShiptoCity = "Duluth";
                    order.ShiptoCompany = "Widgets Inc";
                    order.ShiptoCountryCode = "US";
                    order.ShiptoDayPhone = "6785552323";
                    order.ShiptoEveningPhone = "7703334444";
                    order.ShiptoFirstName = "Sally";
                    order.ShiptoLastName = "McGonkyDee";
                    order.ShiptoPostalCode = "30097";
                    order.ShiptoStateRegion = "GA";
                    order.ShiptoTitle = "Director";
                    order.SkipPaymentProcessing = false;
                    order.SpecialInstructions =
                        "Please wrap this in bubble wrap because my FedEx delivery guy is abusive to packages";
                    order.StoreCompleted =
                        false; // this will bypass everything, including shipping. useful only for importing old orders long completed
                    order.StorefrontHostName = "store.mysite.com";
                    order.StoreIfPaymentDeclines =
                        false; // if payment fails, this can send it to Accounts Receivable. Do not want that. Fail if payment fails.
                    order.TaxCounty = "Gwinnett";
                    order.TaxExempt = false;

                    ChannelPartnerOrderTransaction orderTransaction = new ChannelPartnerOrderTransaction();
                    orderTransaction.Successful = false; // we haven't charged the card yet, so this is false.
                    orderTransaction.Details =
                        new List<ChannelPartnerOrderTransactionDetail>(); // we haven't charged the card yet, so this is empty.
                    order.Transaction = orderTransaction;
                    order.TreatWarningsAsErrors = true;

                    var apiResponse = channelPartnerApi.ImportChannelPartnerOrder(order);
                }

                // ---------------------------------------------
                // ---------------------------------------------
                // Example 2 - Order already processed
                // ---------------------------------------------
                // ---------------------------------------------
                
                ChannelPartnerOrder processedOrder = new ChannelPartnerOrder();
                
                // processedOrder.AdvertisingSource = "Friend"; // https://ultracart.atlassian.net/wiki/spaces/ucdoc/pages/1377001/Advertising+Sources
                // processedOrder.AffiliateId = 856234; // https://ultracart.atlassian.net/wiki/spaces/ucdoc/pages/1377727/Affiliates
                // processedOrder.AffiliateSubId = 1234; // https://ultracart.atlassian.net/wiki/spaces/ucdoc/pages/1376754/Allowing+Affiliates+to+use+Sub-IDs
                // processedOrder.ArbitraryShippingHandlingTotal = 9.99;
                // processedOrder.ArbitraryTax = 2.50;
                // processedOrder.ArbitraryTaxRate = 7.0;
                // processedOrder.ArbitraryTaxableSubtotal = 69.99;
                
                processedOrder.AssociateWithCustomerProfileIfPresent = true;
                processedOrder.AutoApprovePurchaseOrder = true;
                processedOrder.BilltoAddress1 = "11460 Johns Creek Parkway";
                processedOrder.BilltoAddress2 = "Suite 101";
                processedOrder.BilltoCity = "Duluth";
                processedOrder.BilltoCompany = "Widgets Inc";
                processedOrder.BilltoCountryCode = "US";
                processedOrder.BilltoDayPhone = "6784153823";
                processedOrder.BilltoEveningPhone = "6784154019";
                processedOrder.BilltoFirstName = "John";
                processedOrder.BilltoLastName = "Smith";
                processedOrder.BilltoPostalCode = "30097";
                processedOrder.BilltoStateRegion = "GA";
                processedOrder.BilltoTitle = "Sir";
                processedOrder.CcEmail = "orders@widgets.com";
                processedOrder.ChannelPartnerOrderId = "widget-1245-abc";
                processedOrder.ConsiderRecurring = false;
                processedOrder.Coupons = new List<string> { "10OFF", "BUY1GET1" };
                
                // processedOrder.CreditCardAuthorizationAmount = 69.99;
                // processedOrder.CreditCardAuthorizationDts = DateTime.Now.ToString("yyyy-MM-ddTHH:mm:ssK"); // this will usually not be 'now'. it will be the actual auth date
                // processedOrder.CreditCardAuthorizationNumber = "1234";
                
                processedOrder.CreditCardExpirationMonth = 5;
                processedOrder.CreditCardExpirationYear = 2032;
                processedOrder.CreditCardType = "VISA"; // see the hosted fields below for the card number and cvv tokens
                processedOrder.CustomField1 = "Whatever";
                processedOrder.CustomField2 = "You";
                processedOrder.CustomField3 = "Want";
                processedOrder.CustomField4 = "Can";
                processedOrder.CustomField5 = "Go";
                processedOrder.CustomField6 = "In";
                processedOrder.CustomField7 = "CustomFields";
                processedOrder.DeliveryDate = DateTime.Now.ToString("yyyy-MM-ddTHH:mm:ssK");
                processedOrder.Email = "ceo@widgets.com";
                processedOrder.Gift = false;
                
                processedOrder.GiftEmail = "sally@aol.com";
                processedOrder.GiftMessage = "Congratulations on your promotion!";
                
                // processedOrder.InsuranceApplicationId = insuranceApplicationId; // these are used by only a handful of specialized merchants
                // processedOrder.InsuranceClaimId = insuranceClaimId; // these are used by only a handful of specialized merchants
                
                processedOrder.IpAddress = "34.125.95.217";
                
                // -- Items start ---
                ChannelPartnerOrderItem processedItem = new ChannelPartnerOrderItem();
                // processedItem.ArbitraryUnitCost = 9.99;
                // processedItem.AutoOrderLastRebillDts = DateTime.Now.ToString("yyyy-MM-ddTHH:mm:ssK");
                // processedItem.AutoOrderSchedule = "Weekly";
                
                processedItem.MerchantItemId = "shirt";
                processedItem.Quantity = 1;
                processedItem.Upsell = false;
                
                ChannelPartnerOrderItemOption processedItemOption1 = new ChannelPartnerOrderItemOption();
                processedItemOption1.Name = "Size";
                processedItemOption1.Value = "Small";
                
                ChannelPartnerOrderItemOption processedItemOption2 = new ChannelPartnerOrderItemOption();
                processedItemOption2.Name = "Color";
                processedItemOption2.Value = "Orange";
                
                processedItem.Options = new List<ChannelPartnerOrderItemOption> { processedItemOption1, processedItemOption2 };
                
                processedOrder.Items = new List<ChannelPartnerOrderItem> { processedItem };
                // -- Items End ---
                
                // We don't use least cost routing here. We've already completed the order and should know what shipping method
                // the customer was charged for. So that is set in the processedOrder.ShippingMethod property.
                // processedOrder.LeastCostRoute = true; // Give me the lowest cost shipping
                // processedOrder.LeastCostRouteShippingMethods = new List<string> { "FedEx: Ground", "UPS: Ground", "USPS: Priority" };
                processedOrder.MailingListOptIn = true; // Yes, I confirmed with the customer personally they wish to be on my mailing lists.
                processedOrder.NoRealtimePaymentProcessing = true; // nothing to charge, payment was already collected
                processedOrder.PaymentMethod = ChannelPartnerOrder.PaymentMethodEnum.CreditCard;
                // processedOrder.PurchaseOrderNumber = "PO-12345";
                processedOrder.RotatingTransactionGatewayCode = "MyStripe"; // We wish this to be charged against our Stripe gateway
                processedOrder.ScreenBrandingThemeCode = "SF1986"; // Theme codes predated StoreFronts. Each StoreFront still has a theme code under the hood. We need that here. See this screen to find your code: https://secure.ultracart.com/merchant/configuration/customerServiceLoad.do
                processedOrder.ShipOnDate = DateTime.Now.ToString("yyyy-MM-ddTHH:mm:ssK");
                processedOrder.ShipToResidential = true;
                processedOrder.ShippingMethod = "FedEx: Ground";
                processedOrder.ShiptoAddress1 = "55 Main Street";
                processedOrder.ShiptoAddress2 = "Suite 202";
                processedOrder.ShiptoCity = "Duluth";
                processedOrder.ShiptoCompany = "Widgets Inc";
                processedOrder.ShiptoCountryCode = "US";
                processedOrder.ShiptoDayPhone = "6785552323";
                processedOrder.ShiptoEveningPhone = "7703334444";
                processedOrder.ShiptoFirstName = "Sally";
                processedOrder.ShiptoLastName = "McGonkyDee";
                processedOrder.ShiptoPostalCode = "30097";
                processedOrder.ShiptoStateRegion = "GA";
                processedOrder.ShiptoTitle = "Director";
                processedOrder.SkipPaymentProcessing = true; // bypass payment
                processedOrder.SpecialInstructions = "Please wrap this in bubble wrap because my FedEx delivery guy is abusive to packages";
                processedOrder.StoreCompleted = true; // this is an old order or an order handled completely outside UltraCart, so do not do anything to it. Just store it.
                processedOrder.StorefrontHostName = "store.mysite.com";
                processedOrder.StoreIfPaymentDeclines = false; // if payment fails, this can send it to Accounts Receivable. Do not want that. Fail if payment fails.
                processedOrder.TaxCounty = "Gwinnett";
                processedOrder.TaxExempt = false;
                
                ChannelPartnerOrderTransaction processedOrderTransaction = new ChannelPartnerOrderTransaction();
                processedOrderTransaction.Successful = true; // transaction was successful
                
                ChannelPartnerOrderTransactionDetail td1 = new ChannelPartnerOrderTransactionDetail(); // 'td' is short for transaction detail
                td1.Name = "AVS Code";
                td1.Value = "X";
                
                ChannelPartnerOrderTransactionDetail td2 = new ChannelPartnerOrderTransactionDetail();
                td2.Name = "Authorization Code";
                td2.Value = "123456";
                
                ChannelPartnerOrderTransactionDetail td3 = new ChannelPartnerOrderTransactionDetail();
                td3.Name = "CVV Code";
                td3.Value = "M";
                
                ChannelPartnerOrderTransactionDetail td4 = new ChannelPartnerOrderTransactionDetail();
                td4.Name = "Response Code";
                td4.Value = "Authorized";
                
                ChannelPartnerOrderTransactionDetail td5 = new ChannelPartnerOrderTransactionDetail();
                td5.Name = "Reason Code";
                td5.Value = "1";
                
                ChannelPartnerOrderTransactionDetail td6 = new ChannelPartnerOrderTransactionDetail();
                td6.Name = "Response Subcode";
                td6.Value = "1";
                
                ChannelPartnerOrderTransactionDetail td7 = new ChannelPartnerOrderTransactionDetail();
                td7.Name = "Transaction ID";
                td7.Value = "1234567890";
                
                processedOrderTransaction.Details = new List<ChannelPartnerOrderTransactionDetail> { td1, td2, td3, td4, td5, td6, td7 };
                processedOrder.Transaction = processedOrderTransaction;
                processedOrder.TreatWarningsAsErrors = true;
                
                var processedApiResponse = channelPartnerApi.ImportChannelPartnerOrder(processedOrder);
                
                Console.WriteLine("Orders imported successfully");
            }
            catch (Exception ex)
            {
                Console.WriteLine($"Error: {ex.Message}");
                Console.WriteLine(ex.StackTrace);
            }
        }
    }
}

File: ruby/channel_partner/import_channel_partner_order.rb
# To run channel partner examples, you will need:
# 1) An API Key: https://secure.ultracart.com/merchant/configuration/apiManagementApp.do
# 2) That API Key must be assigned to a channel partner: https://secure.ultracart.com/merchant/configuration/customChannelPartnerListLoad.do
#
# The spreadsheet import docs will serve you well here. They describe many fields
# https://ultracart.atlassian.net/wiki/spaces/ucdoc/pages/1377246/Channel+Partner+API+-+Spreadsheet+Import

require 'ultracart_api'
require_relative '../constants'

channel_partner_api = UltracartClient::ChannelPartnerApi.new_using_api_key(Constants::CHANNEL_PARTNER_API_KEY)

# NOTICE:
# The real difficulty with using this API is the hosted fields requirement for credit card information.
# You will need to incorporate UltraCart hosted fields in your Customer Service UI and capture credit card
# information through the hosted fields and then provide the tokens here. You CANNOT provide raw credit
# card information via this interface.
# The two fields in this API are hostedFieldsCardToken and hostedFieldsCvvToken
# Within this sdk_samples github project, review the /hosted_fields/hosted_fields.html file for an example

# ---------------------------------------------
# ---------------------------------------------
# Example 1 - Order needs payment processing
# ---------------------------------------------
# ---------------------------------------------

order = UltracartClient::ChannelPartnerOrder.new

# order.advertising_source = "Friend" # https://ultracart.atlassian.net/wiki/spaces/ucdoc/pages/1377001/Advertising+Sources
# order.affiliate_id = 856234 # https://ultracart.atlassian.net/wiki/spaces/ucdoc/pages/1377727/Affiliates
# order.affiliate_sub_id = 1234 # https://ultracart.atlassian.net/wiki/spaces/ucdoc/pages/1376754/Allowing+Affiliates+to+use+Sub-IDs
# order.arbitrary_shipping_handling_total = 9.99
# order.arbitrary_tax = 2.50
# order.arbitrary_tax_rate = 7.0
# order.arbitrary_taxable_subtotal = 69.99

order.associate_with_customer_profile_if_present = true
order.auto_approve_purchase_order = true
order.billto_address1 = "11460 Johns Creek Parkway"
order.billto_address2 = "Suite 101"
order.billto_city = "Duluth"
order.billto_company = "Widgets Inc"
order.billto_country_code = "US"
order.billto_day_phone = "6784153823"
order.billto_evening_phone = "6784154019"
order.billto_first_name = "John"
order.billto_last_name = "Smith"
order.billto_postal_code = "30097"
order.billto_state_region = "GA"
order.billto_title = "Sir"
order.cc_email = "orders@widgets.com"
order.channel_partner_order_id = "widget-1245-abc"
order.consider_recurring = false
order.coupons = ["10OFF", "BUY1GET1"]

order.credit_card_expiration_month = 5
order.credit_card_expiration_year = 2032
order.credit_card_type = "VISA" # see the hosted fields below for the card number and cvv tokens
order.custom_field1 = "Whatever"
order.custom_field2 = "You"
order.custom_field3 = "Want"
order.custom_field4 = "Can"
order.custom_field5 = "Go"
order.custom_field6 = "In"
order.custom_field7 = "CustomFields"
order.delivery_date = Time.now.strftime('%Y-%m-%dT00:00:00+00:00')
order.email = "ceo@widgets.com"
order.gift = false

order.gift_email = "sally@aol.com"
order.gift_message = "Congratulations on your promotion!"

order.hosted_fields_card_token = "7C97B0AAA26AB10180B4B29F00380101"
order.hosted_fields_cvv_token = "C684AB4336787F0180B4B51971380101"

order.ip_address = "34.125.95.217"

# -- Items start ---
item = UltracartClient::ChannelPartnerOrderItem.new
# item.arbitrary_unit_cost = 9.99
# item.auto_order_last_rebill_dts = Time.now.strftime('%Y-%m-%dT00:00:00+00:00')
# item.auto_order_schedule = "Weekly"

item.merchant_item_id = "shirt"
item.quantity = 1
item.upsell = false

item_option1 = UltracartClient::ChannelPartnerOrderItemOption.new
item_option1.name = "Size"
item_option1.value = "Small"

item_option2 = UltracartClient::ChannelPartnerOrderItemOption.new
item_option2.name = "Color"
item_option2.value = "Orange"

item.options = [item_option1, item_option2]

order.items = [item]
# -- Items End ---

order.least_cost_route = true # Give me the lowest cost shipping
order.least_cost_route_shipping_methods = ["FedEx: Ground", "UPS: Ground", "USPS: Priority"]
order.mailing_list_opt_in = true # Yes, I confirmed with the customer personally they wish to be on my mailing lists.
order.no_realtime_payment_processing = false
order.payment_method = "Credit Card"
order.rotating_transaction_gateway_code = "MyStripe" # We wish this to be charged against our Stripe gateway
order.screen_branding_theme_code = "SF1986" # Theme codes predated StoreFronts. Each StoreFront still has a theme code under the hood
order.ship_on_date = Time.now.strftime('%Y-%m-%dT00:00:00+00:00')
order.ship_to_residential = true
order.shipto_address1 = "55 Main Street"
order.shipto_address2 = "Suite 202"
order.shipto_city = "Duluth"
order.shipto_company = "Widgets Inc"
order.shipto_country_code = "US"
order.shipto_day_phone = "6785552323"
order.shipto_evening_phone = "7703334444"
order.shipto_first_name = "Sally"
order.shipto_last_name = "McGonkyDee"
order.shipto_postal_code = "30097"
order.shipto_state_region = "GA"
order.shipto_title = "Director"
order.skip_payment_processing = false
order.special_instructions = "Please wrap this in bubble wrap because my FedEx delivery guy is abusive to packages"
order.store_completed = false # this will bypass everything, including shipping. useful only for importing old orders long completed
order.storefront_host_name = 'store.mysite.com'
order.store_if_payment_declines = false # if payment fails, this can send it to Accounts Receivable. Do not want that. Fail if payment fails.
order.tax_county = "Gwinnett"
order.tax_exempt = false

order_transaction = UltracartClient::ChannelPartnerOrderTransaction.new
order_transaction.successful = false # we haven't charged the card yet, so this is false
order_transaction.details = [] # we haven't charged the card yet, so this is empty
order.transaction = order_transaction
order.treat_warnings_as_errors = true

api_response = channel_partner_api.import_channel_partner_order(order)

# ---------------------------------------------
# ---------------------------------------------
# Example 2 - Order already processed
# ---------------------------------------------
# ---------------------------------------------

order = UltracartClient::ChannelPartnerOrder.new

order.associate_with_customer_profile_if_present = true
order.auto_approve_purchase_order = true
order.billto_address1 = "11460 Johns Creek Parkway"
order.billto_address2 = "Suite 101"
order.billto_city = "Duluth"
order.billto_company = "Widgets Inc"
order.billto_country_code = "US"
order.billto_day_phone = "6784153823"
order.billto_evening_phone = "6784154019"
order.billto_first_name = "John"
order.billto_last_name = "Smith"
order.billto_postal_code = "30097"
order.billto_state_region = "GA"
order.billto_title = "Sir"
order.cc_email = "orders@widgets.com"
order.channel_partner_order_id = "widget-1245-abc"
order.consider_recurring = false
order.coupons = ["10OFF", "BUY1GET1"]

order.credit_card_expiration_month = 5
order.credit_card_expiration_year = 2032
order.credit_card_type = "VISA"
order.custom_field1 = "Whatever"
order.custom_field2 = "You"
order.custom_field3 = "Want"
order.custom_field4 = "Can"
order.custom_field5 = "Go"
order.custom_field6 = "In"
order.custom_field7 = "CustomFields"
order.delivery_date = Time.now.strftime('%Y-%m-%dT00:00:00+00:00')
order.email = "ceo@widgets.com"
order.gift = false

order.gift_email = "sally@aol.com"
order.gift_message = "Congratulations on your promotion!"

order.ip_address = "34.125.95.217"

# -- Items start ---
item = UltracartClient::ChannelPartnerOrderItem.new

item.merchant_item_id = "shirt"
item.quantity = 1
item.upsell = false

item_option1 = UltracartClient::ChannelPartnerOrderItemOption.new
item_option1.name = "Size"
item_option1.value = "Small"

item_option2 = UltracartClient::ChannelPartnerOrderItemOption.new
item_option2.name = "Color"
item_option2.value = "Orange"

item.options = [item_option1, item_option2]

order.items = [item]
# -- Items End ---

order.mailing_list_opt_in = true # Yes, I confirmed with the customer personally they wish to be on my mailing lists.
order.no_realtime_payment_processing = true # nothing to charge, payment was already collected
order.payment_method = "Credit Card"
order.rotating_transaction_gateway_code = "MyStripe" # We wish this to be charged against our Stripe gateway
order.screen_branding_theme_code = "SF1986" # Theme codes predated StoreFronts. Each StoreFront still has a theme code under the hood
order.ship_on_date = Time.now.strftime('%Y-%m-%dT00:00:00+00:00')
order.ship_to_residential = true
order.shipping_method = "FedEx: Ground"
order.shipto_address1 = "55 Main Street"
order.shipto_address2 = "Suite 202"
order.shipto_city = "Duluth"
order.shipto_company = "Widgets Inc"
order.shipto_country_code = "US"
order.shipto_day_phone = "6785552323"
order.shipto_evening_phone = "7703334444"
order.shipto_first_name = "Sally"
order.shipto_last_name = "McGonkyDee"
order.shipto_postal_code = "30097"
order.shipto_state_region = "GA"
order.shipto_title = "Director"
order.skip_payment_processing = true # bypass payment
order.special_instructions = "Please wrap this in bubble wrap because my FedEx delivery guy is abusive to packages"
order.store_completed = true # this is an old order or an order handled completely outside UltraCart, so do not do anything to it. Just store it.
order.storefront_host_name = 'store.mysite.com'
order.store_if_payment_declines = false # if payment fails, this can send it to Accounts Receivable. Do not want that. Fail if payment fails.
order.tax_county = "Gwinnett"
order.tax_exempt = false

order_transaction = UltracartClient::ChannelPartnerOrderTransaction.new
order_transaction.successful = true

td1 = UltracartClient::ChannelPartnerOrderTransactionDetail.new
td1.name = "AVS Code"
td1.value = "X"

td2 = UltracartClient::ChannelPartnerOrderTransactionDetail.new
td2.name = "Authorization Code"
td2.value = "123456"

td3 = UltracartClient::ChannelPartnerOrderTransactionDetail.new
td3.name = "CVV Code"
td3.value = "M"

td4 = UltracartClient::ChannelPartnerOrderTransactionDetail.new
td4.name = "Response Code"
td4.value = "Authorized"

td5 = UltracartClient::ChannelPartnerOrderTransactionDetail.new
td5.name = "Reason Code"
td5.value = "1"

td6 = UltracartClient::ChannelPartnerOrderTransactionDetail.new
td6.name = "Response Subcode"
td6.value = "1"

td7 = UltracartClient::ChannelPartnerOrderTransactionDetail.new
td7.name = "Transaction ID"
td7.value = "1234567890"

order_transaction.details = [td1, td2, td3, td4, td5, td6, td7]
order.transaction = order_transaction
order.treat_warnings_as_errors = true

api_response = channel_partner_api.import_channel_partner_order(order)